### PR TITLE
Resiliency Support for Bulk Subscribe

### PR DESCRIPTION
**Pull request by DeepanshuA**
Signed-off-by: Deepanshu Agarwal <deepanshu.agarwal1984@gmail.com>

# Description

Till now, for bulk subscribe, all messages would be replayed under resiliency policy, irrespective of whether they passed earlier or not. Whereas, a Filter should be applied so that successful entries are not replayed.

## Issue reference

Please reference the issue this PR will close: #5501 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: utils/utils.go
Comment (mukundansundar) [inline utils/utils.go:174]:
```suggestion
	filteredItems := make([]T, 0, len(items))
	for _, value := range items {
		if fn(value) {
			filteredItems = append(filteredItems, value)
		}
	}
	return filteredItems[:len(filteredItems):len(filteredItems)] // return and trunc capacity of slice
```

File: utils/utils.go
Comment (DeepanshuA) [reply utils/utils.go:165]: Sure, that actually makes sense. What I had started to think was someone using `Filter` and forgetting about it, with some specific flags - but then instead of setting/re-setting the flag, they can chose to call/not-call Filter itself.

File: pkg/runtime/bulk_subscriber.go
Comment (ItalyPaleAle) [inline pkg/runtime/bulk_subscriber.go:12]: Please use sync/atomic instead

File: utils/utils.go
Comment (shubham1172) [reply utils/utils.go:174]: `return filteredItems[:len(filteredItems)] // return the truncated slice`

File: utils/utils_test.go
Comment (shubham1172) [inline utils/utils_test.go:243]: Please also add a test for capacity `cap`.

File: pkg/runtime/bulk_subscriber.go
Comment (shubham1172) [inline pkg/runtime/bulk_subscriber.go:288]: This can be simplified to a `Store` instead of `CompareAndSwap`, right?

File: pkg/runtime/bulk_subscriber.go
Comment (shubham1172) [inline pkg/runtime/bulk_subscriber.go:284]:
```suggestion
			}, false)
```

Or, rather we will remove this since the `Filter` method is changing.

File: pkg/runtime/bulk_subscriber.go
Comment (ItalyPaleAle) [reply pkg/runtime/bulk_subscriber.go:288]: CAS is correct to use in this case to ensure the operation is actually atomic, but it shouldn't be done here because this can lead to race conditions: while operations like `Load` are performed atomically, there's no lock around them. CAS returns a boolean indicating if the swap happened and you can use that instead:

```go
// If the swap happened, returns true
if firstAttempt.CompareAndSwap(true, false) {
	newPubSubMsg := utils.Filter(psm.pubSubMessages, func(ps pubSubMessage) bool {
		index := (*entryIdIndexMap)[ps.entry.EntryId]
		return (*bulkResponses)[index].Error != nil
	}, false) // We can put false directly because we know it's false in this block
	psm.pubSubMessages = newPubSubMsg
	psm.length = len(newPubSubMsg)
}
```

File: pkg/runtime/bulk_subscriber.go
Comment (shubham1172) [reply pkg/runtime/bulk_subscriber.go:360]: Haha, I saw you refactoring other instances of `fmt.Sprintf` to `+` as well. Is it purely due to perf. reasons?

image

[[source](https://forum.golangbridge.org/t/fmt-sprintf-vs-string-concatenation/23006)]

Comment (ItalyPaleAle) [reply pkg/runtime/bulk_subscriber.go:360]: Yes. It's consistent with benchmarks I ran too in another PR. When it's just about string concatenation in a hot path, the perf difference is significant. Not just the ns/op, but the fact that the second has 0 heap allocations, while the first one has a number (in my test, I got 3) - and that's more work for the allocator and the GC.

Comment (shubham1172) [reply pkg/runtime/bulk_subscriber.go:360]: Good stuff! I wonder if it's worth adding it to our linter because it can get tedious to manually review it each time.

Comment (ItalyPaleAle) [reply pkg/runtime/bulk_subscriber.go:360]: I just looked but couldn't see an option in our linters to catch these kinds of things. I don't think we should forbid `fmt.Sprintf` entirely because there are many situations where it's fine: in tests and in other non-hot paths, and/or when there are very complex concatenations (such as where the benefits in terms of code readability outweight the perf impact)

Comment (shubham1172) [reply pkg/runtime/bulk_subscriber.go:360]: Yes, if we ever decide on configuring our linter, it should only be confined to simple concatenations.

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go:288]: Actually, I had tried to look in that direction when your PR https://github.com/dapr/dapr/pull/5538 was raised, but issue is that in case of these Bulk APIs, we want to keep trying (say retries) decrementally i.e. omitting out all Successful entries and just want to keep re-try for Failed entries. And, that all becomes a part of Policy Only, But if we return from policy, we get outside Policy. 
How can we refactor so that we use return of `policy` but also remain inside it?

Comment (ItalyPaleAle) [reply pkg/runtime/bulk_subscriber.go:288]: Ok, I see what you're trying to do now. I don't think this can be done without side effects, indeed. You should be able to make this work safely, even with side effects, by relying on atomic's and CAS. Ideally you will want to use [`atomic.Pointer`](https://pkg.go.dev/sync/atomic#Pointer) to keep a pointer to the last list of objects that need to be requested It's going to be a bit tricky though :)

File: utils/utils.go
Comment (mukundansundar) [reply utils/utils.go:174]: @ItalyPaleAle when i tried with this, I was getting some failures here ... when running multiple unit tests in series, even though the type is only slice and not pointer to slice, changing the values in place was affecting the source !

File: utils/utils.go
Comment (DeepanshuA) [reply utils/utils.go]: Done.

File: utils/utils.go
Comment (DeepanshuA) [reply utils/utils.go]: Removed the flag.

File: utils/utils.go
Comment (DeepanshuA) [reply utils/utils.go]: Used array with a capacity pre-defined. But, as discussed, not doing in-place.

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go]: Done.

File: utils/utils.go
Comment (DeepanshuA) [reply utils/utils.go]: Done

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go]: As discussed with @ItalyPaleAle and @mukundansundar, changed the implementation. Using Accumulator now from NewRunnerWithOptions.

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go]: Yup, removed.

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go]: Changed.

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go]: Done

File: pkg/runtime/bulksub_resiliency.go
Comment (DeepanshuA) [reply pkg/runtime/bulksub_resiliency.go:39]: 1. Since subscribedEntries load psm.pubSubMessages (`subscribedEntries.Store(&psm.pubSubMessages)`), which have already changed as in line above, so modified psm is loaded, instead of original psm. And, it is inside Accumulator, so it seems safe.
2. I am just thinking, if atomic.Pointer is still needed, since these modifications are getting done only inside Accumulator now.

**Comment by DeepanshuA** [comment]
@ItalyPaleAle @shubham1172 @mukundansundar @dapr/approvers-dapr @dapr/maintainers-dapr Re-requesting review here. Whenever I re-request review from someone, some other re-review request gets removed, so requesting via comment.

File: utils/utils_test.go
Comment (shubham1172) [reply utils/utils_test.go:243]: @DeepanshuA

File: pkg/runtime/runtime.go
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:854]: Isn't this the same as policyDef (from L730)?

File: pkg/runtime/bulksub_resiliency.go
Comment (ItalyPaleAle) [inline pkg/runtime/bulksub_resiliency.go:51]: Nit: you may want to `panic` here as this should never happen, and if it happens, it should be caught during development

File: pkg/runtime/bulksub_resiliency.go
Comment (ItalyPaleAle) [inline pkg/runtime/bulksub_resiliency.go:30]: Is this guaranteed to be found, or is there a chance that index will be the zero value and so `(*bscData.bulkResponses)[index]` would be nil? should there be a nil-check?

File: pkg/runtime/bulksub_resiliency.go
Comment (ItalyPaleAle) [inline pkg/runtime/bulksub_resiliency.go:24]: Why is this using atomic? I see it's only used in the accumulator, so it shouldn't need to be atomic anymore, unless I misunderstand

**Comment by ItalyPaleAle** [comment]
@DeepanshuA  FYI I have merged changes from master into this PR. Because of #5648, the `invokev1` package changed a lot and didn't want to put the burden on you to fix the merge conflicts :) (and it also helped me find a bug!)

File: pkg/runtime/runtime.go
Comment (DeepanshuA) [reply pkg/runtime/runtime.go:854]: Yeah, correct, I will use the already defined policyDef here.

File: pkg/runtime/bulksub_resiliency.go
Comment (DeepanshuA) [reply pkg/runtime/bulksub_resiliency.go:51]: Sure, changing that in bulk Subscribe, though not touching same logic in runtime `subscribe` functionality and letting that remain same.

File: pkg/runtime/bulksub_resiliency.go
Comment (DeepanshuA) [reply pkg/runtime/bulksub_resiliency.go:30]: Yeah, this is guaranteed to be found. But, still it doesn't harm to put an extra check here, so putting an explicit check here as well.

File: pkg/runtime/bulksub_resiliency.go
Comment (DeepanshuA) [reply pkg/runtime/bulksub_resiliency.go:24]: Yeah, correct - atomic shouldn't be required - changing that.

File: utils/utils_test.go
Comment (DeepanshuA) [reply utils/utils_test.go:243]: I have added cap assertions in the test cases, please see if that helps.

Comment (shubham1172) [reply utils/utils_test.go:243]: Looks good, thanks @DeepanshuA!

**Comment by DeepanshuA** [comment]
go test -race results:

/usr/local/go/bin/go test -race -run ^TestBulkSubscribeResiliency$ github.com/dapr/dapr/pkg/runtime
ok      github.com/dapr/dapr/pkg/runtime        28.265s

/usr/local/go/bin/go test -race -run ^TestBulkSubscribeResiliencyWithLongRetries$ github.com/dapr/dapr/pkg/runtime
ok      github.com/dapr/dapr/pkg/runtime        12.202s

/usr/local/go/bin/go test -race -run ^TestBulkSubscribeResiliencyStateConversionsFromHalfOpen$ github.com/dapr/dapr/pkg/runtime
ok      github.com/dapr/dapr/pkg/runtime        13.232s

File: pkg/runtime/bulk_subscriber.go
Comment (mukundansundar) [inline pkg/runtime/bulk_subscriber.go:472]: extract as error var and reuse in line below.

File: pkg/runtime/bulk_subscriber.go
Comment (mukundansundar) [inline pkg/runtime/bulk_subscriber.go:702]: can there be scenario that the entryId is not present in the map?

Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go:702]: No, the way it can reach this code-piece, entryIdIndexMap will have those entries.
That is taken care of separately when entries are sent by App, and if they are outside the known entries.

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go]: Done.

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go]: Thanks, useful info. Changed accordingly.

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go:87]: Changed comments as required.

File: pkg/runtime/bulk_subscriber.go
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go:35]: Provided.

**Review by artursouza — APPROVED** [review]

**Comment by artursouza** [comment]
Will merge after @ItalyPaleAle approves.

**Review by ItalyPaleAle — APPROVED** [review]
LGTM

**Pull request merged**
